README.md
[![DOI](https://zenodo.org/badge/217881799.svg)](https://zenodo.org/badge/latestdoi/217881799)

## Weighted boxes fusion

Repository contains Python implementation of several methods for ensembling boxes from object detection models: 

* Non-maximum Suppression (NMS)
* Soft-NMS [[1]](https://arxiv.org/abs/1704.04503)
* Non-maximum weighted (NMW) [[2]](http://openaccess.thecvf.com/content_ICCV_2017_workshops/papers/w14/Zhou_CAD_Scale_Invariant_ICCV_2017_paper.pdf)
* **Weighted boxes fusion (WBF)** [[3]](https://arxiv.org/abs/1910.13302) - new method which gives better results comparing to others 

## Requirements

Python 3.*, Numpy, Numba

# Installation

`pip install ensemble-boxes`

## Usage examples

Coordinates for boxes expected to be normalized e.g in range [0; 1]. Order: x1, y1, x2, y2. 

Example of boxes ensembling for 2 models below. 
* First model predicts 5 boxes, second model predicts 4 boxes.
* Confidence scores for each box model 1: [0.9, 0.8, 0.2, 0.4, 0.7]
* Confidence scores for each box model 2: [0.5, 0.8, 0.7, 0.3]
* Labels (classes) for each box model 1: [0, 1, 0, 1, 1]
* Labels (classes) for each box model 2: [1, 1, 1, 0]
* We set weight for 1st model to be 2, and weight for second model to be 1.
* We set intersection over union for boxes to be match: iou_thr = 0.5
* We skip boxes with confidence lower than skip_box_thr = 0.0001

```python
from ensemble_boxes import *

boxes_list = [[
    [0.00, 0.51, 0.81, 0.91],
    [0.10, 0.31, 0.71, 0.61],
    [0.01, 0.32, 0.83, 0.93],
    [0.02, 0.53, 0.11, 0.94],
    [0.03, 0.24, 0.12, 0.35],
],[
    [0.04, 0.56, 0.84, 0.92],
    [0.12, 0.33, 0.72, 0.64],
    [0.38, 0.66, 0.79, 0.95],
    [0.08, 0.49, 0.21, 0.89],
]]
scores_list = [[0.9, 0.8, 0.2, 0.4, 0.7], [0.5, 0.8, 0.7, 0.3]]
labels_list = [[0, 1, 0, 1, 1], [1, 1, 1, 0]]
weights = [2, 1]

iou_thr = 0.5
skip_box_thr = 0.0001
sigma = 0.1

boxes, scores, labels = nms(boxes_list, scores_list, labels_list, weights=weights, iou_thr=iou_thr)
boxes, scores, labels = soft_nms(boxes_list, scores_list, labels_list, weights=weights, iou_thr=iou_thr, sigma=sigma, thresh=skip_box_thr)
boxes, scores, labels = non_maximum_weighted(boxes_list, scores_list, labels_list, weights=weights, iou_thr=iou_thr, skip_box_thr=skip_box_thr)
boxes, scores, labels = weighted_boxes_fusion(boxes_list, scores_list, labels_list, weights=weights, iou_thr=iou_thr, skip_box_thr=skip_box_thr)
```

#### Single model

If you need to apply NMS or any other method to single model predictions you can call function like that:

```python
from ensemble_boxes import *
# Merge boxes for single model predictions
boxes, scores, labels = weighted_boxes_fusion([boxes_list], [scores_list], [labels_list], weights=None, method=method, iou_thr=iou_thr, thresh=thresh)
```

More examples can be found in [example.py](examples/example.py)

#### 3D version

There is support for 3D boxes in WBF method with `weighted_boxes_fusion_3d` function. Check example of usage in [example_3d.py](examples/example_3d.py)

#### 1D version

There is support for 1D line segments in WBF method with `weighted_boxes_fusion_1d` function. Check example of usage in [example_1d.py](examples/example_1d.py). It was reported that 1D variant can be useful in Named-entity recognition (NER) type of tasks for Natural Language Processing (NLP) problems. Check discussion [here](https://www.kaggle.com/c/feedback-prize-2021/discussion/313389).

## Benchmarks

* Benchmark for [Open Images Dataset (5 models)](benchmark_oid/README.md)
* Benchmark for [COCO Dataset (10 models)](benchmark_coco/README.md)
* Benchmark for [NLP Dataset (10 models)](benchmark_nlp/README.md) - example for one-dimensional WBF variant

## Description of WBF method and citation

* https://arxiv.org/abs/1910.13302 (updated: 2020.08)
* https://authors.elsevier.com/c/1ca0dxnVK3cWY 

If you find this code useful please cite:

```
@article{solovyev2021weighted,
  title={Weighted boxes fusion: Ensembling boxes from different object detection models},
  author={Solovyev, Roman and Wang, Weimin and Gabruseva, Tatiana},
  journal={Image and Vision Computing},
  pages={1-6},
  year={2021},
  publisher={Elsevier}
}
```


CHANGES.md
# Release notes
All notable changes to this project will be documented in this file.

##  v1.0.9
- WBF 1D variant for line segments was added. It was reported that 1D variant can be useful in Named-entity recognition (NER) type of tasks for Natural Language Processing (NLP) problems. Check discussion [here](https://www.kaggle.com/c/feedback-prize-2021/discussion/313389).
- Small comments/syntax fixes, removed unused functions, removed unused numba dependency. Slightly increased speed for default 'avg' method.
- Added new version of WBF, which works faster because of more vectorized structure. It's avoiding cycles and if-statements. Changes are allowed to increase speed 20-30%. Score on test data a little bit decreased from 0.598214 to 0.597297.
To use: ```from ensemble_boxes.ensemble_boxes_wbf_experimental import weighted_boxes_fusion_experimental```
- NLP benchmark was added. It's used for WBF 1D variant. Credits for benchmarks and idea of using WBF for NLP task goes to [Chris Deotte](https://www.kaggle.com/cdeotte) and [Udbhav Bamba](https://www.kaggle.com/ubamba98).
- Benchmarks moved into separate folders (benchmark_oid, benchmark_coco, benchmark_nlp)

##  v1.0.8
- Speed up of find_matching_box function. See details [here](https://github.com/ZFTurbo/Weighted-Boxes-Fusion/pull/48). 
  OID benchmark: 285 sec -> 242 sec. COCO benchmark: 1055 sec -> 643 sec

##  v1.0.7
- Fixed incorrect values after WBF for allows_overflow = False mode. See details [here](https://github.com/ZFTurbo/Weighted-Boxes-Fusion/pull/41).
- Fixed incorrect values after WBF for conf_type = 'max' mode. See details [here](https://github.com/ZFTurbo/Weighted-Boxes-Fusion/pull/42). 

##  v1.0.6
- Added 2 new methods for average boxes: '_box_and_model_avg_' and '_absent_model_aware_avg_'. See details [here](https://github.com/ZFTurbo/Weighted-Boxes-Fusion/pull/25).
Both methods fix issue with confidences larger than 1 after ensemble. Also it gives better results for cases when there are more than 1 box goes to cluster from same model.
One of these methods will replace default 'avg' in later releases after proper testing. Thanks [@i-aki-y](https://github.com/i-aki-y) for great PR.
- Added first version of unit tests. 

##  v1.0.5
- Added [benchmark](https://github.com/ZFTurbo/Weighted-Boxes-Fusion/tree/master/benchmark) files

##  v1.0.4
- Added many input data checks in all methods: set to 0.0 coords < 0, set to 1.0 coords > 1, swap cases where x2 < x1 and y2 < y1, remove zero area boxes.
- Added numba @jit(nopython=True) for critical functions (NMS, NMW and WBF). Speed up around x2 times (tested on example_oid.py).
- Added support for 3D boxes with function weighted_boxes_fusion_3d

##  v1.0.1
- Fixed bug with incorrect work of WBF and NMW algorithms if provided boxes was unsorted and skip_box_thr was larger than 0.0.



example_3d.py
# coding: utf-8
__author__ = 'ZFTurbo: https://kaggle.com/zfturbo'


import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection
from ensemble_boxes import *


def plot_cube(ax, cube_definition, lbl, thickness):
    cube_definition_array = [
        np.array(list(item))
        for item in cube_definition
    ]

    points = []
    points += cube_definition_array
    vectors = [
        cube_definition_array[1] - cube_definition_array[0],
        cube_definition_array[2] - cube_definition_array[0],
        cube_definition_array[3] - cube_definition_array[0]
    ]

    points += [cube_definition_array[0] + vectors[0] + vectors[1]]
    points += [cube_definition_array[0] + vectors[0] + vectors[2]]
    points += [cube_definition_array[0] + vectors[1] + vectors[2]]
    points += [cube_definition_array[0] + vectors[0] + vectors[1] + vectors[2]]

    points = np.array(points)

    edges = [
        [points[0], points[3], points[5], points[1]],
        [points[1], points[5], points[7], points[4]],
        [points[4], points[2], points[6], points[7]],
        [points[2], points[6], points[3], points[0]],
        [points[0], points[2], points[4], points[1]],
        [points[3], points[6], points[7], points[5]]
    ]

    faces = Poly3DCollection(edges, linewidths=thickness + 1)
    if lbl == 0:
        faces.set_edgecolor((1, 0, 0))
    else:
        faces.set_edgecolor((0, 0, 1))
    faces.set_facecolor((0, 0, 1, 0.1))

    ax.add_collection3d(faces)
    ax.scatter(points[:, 0], points[:, 1], points[:, 2], s=0)


def show_boxes(boxes_list, scores_list, labels_list, image_size=800):
    image = np.zeros((image_size, image_size, 3), dtype=np.uint8)
    image[...] = 255
    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')

    for i in range(len(boxes_list)):
        for j in range(len(boxes_list[i])):
            x1 = int(image_size * boxes_list[i][j][0])
            y1 = int(image_size * boxes_list[i][j][1])
            z1 = int(image_size * boxes_list[i][j][2])
            x2 = int(image_size * boxes_list[i][j][3])
            y2 = int(image_size * boxes_list[i][j][4])
            z2 = int(image_size * boxes_list[i][j][5])
            lbl = labels_list[i][j]
            cube_definition = [
                (x1, y1, z1), (x1, y2, z1), (x2, y1, z1), (x1, y1, z2)
            ]
            plot_cube(ax, cube_definition, lbl, int(4 * scores_list[i][j]))

    plt.show()


def example_wbf_3d_2_models(iou_thr=0.55, draw_image=True):
    """
    This example shows how to ensemble boxes from 2 models using WBF_3D method
    :return: 
    """

    boxes_list = [
        [
            [0.00, 0.51, 0.41, 0.81, 0.91, 0.78],
            [0.10, 0.31, 0.45, 0.71, 0.61, 0.85],
            [0.01, 0.32, 0.55, 0.83, 0.93, 0.95],
            [0.02, 0.53, 0.11, 0.11, 0.94, 0.55],
            [0.03, 0.24, 0.34, 0.12, 0.35, 0.67],
        ],
        [
            [0.04, 0.56, 0.36, 0.84, 0.92, 0.82],
            [0.12, 0.33, 0.46, 0.72, 0.64, 0.75],
            [0.38, 0.66, 0.55, 0.79, 0.95, 0.90],
            [0.08, 0.49, 0.15, 0.21, 0.89, 0.67],
        ],
    ]
    scores_list = [
        [
            0.9,
            0.8,
            0.2,
            0.4,
            0.7,
        ],
        [
            0.5,
            0.8,
            0.7,
            0.3,
        ]
    ]
    labels_list = [
        [
            0,
            1,
            0,
            1,
            1,
        ],
        [
            1,
            1,
            1,
            0,
        ]
    ]
    weights = [2, 1]
    if draw_image:
        show_boxes(boxes_list, scores_list, labels_list)

    boxes, scores, labels = weighted_boxes_fusion_3d(boxes_list, scores_list, labels_list, weights=weights, iou_thr=iou_thr, skip_box_thr=0.0)

    if draw_image:
        show_boxes([boxes], [scores], [labels.astype(np.int32)])

    print(len(boxes))
    print(boxes)


if __name__ == '__main__':
    draw_image = True
    example_wbf_3d_2_models(iou_thr=0.2, draw_image=draw_image)

